shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/IstioSetupService.java

```diff
@@ -0,0 +1,121 @@
+package com.redhat.service.smartevents.shard.operator;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.redhat.service.smartevents.shard.operator.providers.TemplateProvider;
+import com.redhat.service.smartevents.shard.operator.resources.istio.gateway.Gateway;
+import com.redhat.service.smartevents.shard.operator.resources.istio.requestauthentication.RequestAuthentication;
+import com.redhat.service.smartevents.shard.operator.resources.istio.virtualservice.VirtualService;
+
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.quarkus.runtime.StartupEvent;
+
+@ApplicationScoped
+public class IstioSetupService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(IstioSetupService.class);
+
+    private static final String ISTIO_GATEWAY_NAMESPACE = "knative-eventing";
+    private static final String ISTIO_GATEWAY_NAME = "broker-gateway";
+    private static final String ISTIO_VIRTUAL_SERVICE_NAMESPACE = "knative-eventing";
+    private static final String ISTIO_VIRTUAL_SERVICE_NAME = "broker-virtual-service";
+    private static final String JWT_REQUEST_AUTHENTICATION_NAMESPACE = "istio-system";
+    private static final String JWT_REQUEST_AUTHENTICATION_NAME = "jwt-rh-sso";
+
+    @Inject
+    KubernetesClient kubernetesClient;
+
+    @Inject
+    TemplateProvider templateProvider;
+
+    @ConfigProperty(name = "event-bridge.istio.jwt.issuer")
+    String jwtIssuer;
+
+    @ConfigProperty(name = "event-bridge.istio.jwt.jwksUri")
+    String jwksUri;
+
+    void setupIstioComponent(@Observes StartupEvent event) {
+
+        createIstioGateway();
+
+        createIstioVirtualService();
+
+        createJWTRequestAuthentication();
+    }
+
+    private void createIstioGateway() {
+
+        Gateway existing = kubernetesClient.resources(Gateway.class)
+                .inNamespace(ISTIO_GATEWAY_NAMESPACE)
+                .withName(ISTIO_GATEWAY_NAME)
+                .get();
+
+        if (existing == null) {
+            Gateway expected = templateProvider.loadIstioGatewayTemplate();
+            expected.getMetadata().setName(ISTIO_GATEWAY_NAME);
+            expected.getMetadata().setNamespace(ISTIO_GATEWAY_NAMESPACE);
+
+            try {
+                kubernetesClient.resources(Gateway.class)
+                        .inNamespace(ISTIO_GATEWAY_NAMESPACE)
+                        .create(expected);
```

Comment:
IDK whether we should improve the API use here (and elsewhere).

The `create(..)` method throws an `Exception` **if** the **request** encountered problems.

There is no guarantee the resource has actually been **created** by this method invocation; only that it was **requested**.

We could either poll here until a call to `get()` returns an instance of enhance `BridgeIngressController` to check its Istio depenencies are available... but simply calling `create(...)` in a "fire and forget" approach probably isn't best.

Reply:
That said we never explicitly checked their existence when creating them with `yaml`..

Reply:
AFAIK this is a synchronous operation, so if no exception is thrown then the resource should be created.
Though I am not sure how Istio actually process this resource, whether some service is deployed, or just Istio use it internally.

Reply:
These are Istio internal resource only.

Reply:
I asked the Java Operator SDK team about the `delete(...)` method a while ago; and they categorically advised it only _requests_ deletion and returns if the _request_ was successful. It is then for the applicable Controller to perform the deletion. I'd assume the same applies for `create(..)` i.e. the _request_ is successful and then the applicable Controller performs the provisioning; but if you know otherwise then fair point (although in our tests we call `create(..)` and then poll for things to become available).

Reply:
`delete` is IMHO a different case due to finalizers. So delete request should just set the delete timestamp, the actual deletion is done by k8s based on finalizers content.

Reply:
Great, thanks for this discussion @sutaakar.

I checked the source code and `create(OrReplace)(..)` does indeed return once the item is actually created.

Reply:
Hmmm. I'm still not convinced... but I won't argue my point (much more).

The `create(..)` method asks k8s to provision a resource with the proposed `spec`. A HTTP PUT is invoked to the k8s API Server, that persists the request and returns the requested object. At this point there has been no Controller execute to actually create anything. It is only when the usual CRD reconciliation loop runs is the object created. This is explained well [here](https://medium.com/nerd-for-tech/k8s-deployment-creation-what-happens-behind-the-scenes-4e7362aee295)... although I am extrapolating the article to cover all (Custom) Resources.

Reply:
[This](https://stackoverflow.com/questions/61267334/kubectl-create-apply-synchronously-until-each-object-is-created) seems to confirm my understanding too.